<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Generation of New Variants &#8212; rstoolbox 0.9.3 documentation</title>
    <link rel="stylesheet" href="../_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../_static/style.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/functions.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="../_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="../_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-3.3.7/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="../_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Experimental Data" href="experimental_analysis.html" />
    <link rel="prev" title="Compare Decoy Populations" href="population_comparison.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head><body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="../index.html">
          rstoolbox</a>
        <span class="navbar-text navbar-version pull-left"><b>0.9.3</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="../scripts.html">Scripts</a></li>
                <li><a href="../tutorial.html">Tutorial</a></li>
                <li><a href="../api.html">API</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="../index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">An introduction to rstoolbox</a></li>
<li class="toctree-l1"><a class="reference internal" href="../highlights.html">Highlighted works</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installing.html">Installing and Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../scripts.html">Available direct executables</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../tutorial.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../api.html">API reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../rosettascripts.html">RosettaScripts</a></li>
<li class="toctree-l1"><a class="reference internal" href="../whatsnew.html">What’s new</a></li>
<li class="toctree-l1"><a class="reference internal" href="../contributing.html">Contributing</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">Generation of New Variants</a><ul>
<li><a class="reference internal" href="#loading-a-reference">Loading a Reference</a></li>
<li><a class="reference internal" href="#loading-the-design-data">Loading the Design Data</a></li>
<li><a class="reference internal" href="#minimal-mutant">Minimal Mutant</a></li>
<li><a class="reference internal" href="#defining-new-mutations-of-interest">Defining New Mutations of Interest</a></li>
<li><a class="reference internal" href="#exhaustive-point-mutant-sampling">Exhaustive Point Mutant Sampling</a></li>
<li><a class="reference internal" href="#learning-from-amino-acids-frequency">Learning from Amino Acids Frequency</a></li>
<li><a class="reference internal" href="#creating-the-new-mutants">Creating the New Mutants</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="../search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  <div class="section" id="generation-of-new-variants">
<span id="new-mutants"></span><h1>Generation of New Variants<a class="headerlink" href="#generation-of-new-variants" title="Permalink to this headline">¶</a></h1>
<p>A first round of designs might just be the stepping stone towards a second generation. It can be used to learn and better stir the next generation according to whatever
is our final aim.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">All the examples here will generate new sequences. Once those new sequences are generated, we can generate with a call to <a class="reference internal" href="../generated/DesignFrame/rstoolbox.components.DesignFrame.make_resfile.html#rstoolbox.components.DesignFrame.make_resfile" title="rstoolbox.components.DesignFrame.make_resfile"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DesignFrame.make_resfile()</span></code></a>
the residue files that can be provided to Rosetta through the
<a class="reference external" href="https://www.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/TaskOperations/taskoperations_pages/ReadResfileOperation">ReadResfile</a> to guide
the design process.
We are not calling the method in this tutorial as it generates files which cannot be shown.</p>
</div>
<div class="section" id="loading-a-reference">
<h2>Loading a Reference<a class="headerlink" href="#loading-a-reference" title="Permalink to this headline">¶</a></h2>
<p>As in <a class="reference internal" href="sequence_analysis.html#sequence-analysis"><span class="std std-ref">Sequence Analysis</span></a>, we will need to load a reference with <a class="reference internal" href="../generated/rstoolbox.io.get_sequence_and_structure.html#rstoolbox.io.get_sequence_and_structure" title="rstoolbox.io.get_sequence_and_structure"><code class="xref py py-func docutils literal notranslate"><span class="pre">get_sequence_and_structure()</span></code></a>.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Through all the process several times the <code class="docutils literal notranslate"><span class="pre">chainID</span></code> of the decoy of interest will be called. This is due to the fact that the library can manipulate
decoys with multiple chains (designed or not), and, thus, analysis must be called upon the sequences of interest.</p>
</div>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [1]: </span><span class="kn">import</span> <span class="nn">rstoolbox</span> <span class="kn">as</span> <span class="nn">rs</span>
<span class="gp">   ...: </span><span class="kn">import</span> <span class="nn">pandas</span> <span class="kn">as</span> <span class="nn">pd</span>
<span class="gp">   ...: </span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="kn">as</span> <span class="nn">plt</span>
<span class="gp">   ...: </span><span class="kn">import</span> <span class="nn">seaborn</span> <span class="kn">as</span> <span class="nn">sns</span>
<span class="gp">   ...: </span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.width&#39;</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>
<span class="gp">   ...: </span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span> <span class="mi">500</span><span class="p">)</span>
<span class="gp">   ...: </span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s2">&quot;display.max_seq_items&quot;</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="gp">   ...: </span><span class="n">baseline</span> <span class="o">=</span> <span class="n">rs</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">get_sequence_and_structure</span><span class="p">(</span><span class="s1">&#39;../rstoolbox/tests/data/2pw9C.pdb&#39;</span><span class="p">)</span>
<span class="gp">   ...: </span><span class="n">baseline</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">)</span>
<span class="gp">   ...: </span><span class="n">baseline</span><span class="o">.</span><span class="n">add_reference_sequence</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">baseline</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">))</span>
<span class="gp">   ...: </span><span class="n">baseline</span><span class="o">.</span><span class="n">add_reference_shift</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="mi">32</span><span class="p">)</span>
<span class="gp">   ...: </span>
</pre></div>
</div>
</div>
<div class="section" id="loading-the-design-data">
<h2>Loading the Design Data<a class="headerlink" href="#loading-the-design-data" title="Permalink to this headline">¶</a></h2>
<p>Again, we are mimicking <a class="reference internal" href="sequence_analysis.html#sequence-analysis"><span class="std std-ref">Sequence Analysis</span></a>.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [2]: </span><span class="n">rules</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;scores_ignore&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;fa_*&#39;</span><span class="p">,</span> <span class="s1">&#39;niccd_*&#39;</span><span class="p">,</span> <span class="s1">&#39;hbond_*&#39;</span><span class="p">,</span> <span class="s1">&#39;lk_ball_wtd&#39;</span><span class="p">,</span> <span class="s1">&#39;pro_close&#39;</span><span class="p">,</span> <span class="s1">&#39;dslf_fa13&#39;</span><span class="p">,</span> <span class="s1">&#39;C_ni_rmsd_threshold&#39;</span><span class="p">,</span>
<span class="gp">   ...: </span>                           <span class="s1">&#39;omega&#39;</span><span class="p">,</span> <span class="s1">&#39;p_aa_pp&#39;</span><span class="p">,</span> <span class="s1">&#39;yhh_planarity&#39;</span><span class="p">,</span> <span class="s1">&#39;ref&#39;</span><span class="p">,</span> <span class="s1">&#39;rama_prepro&#39;</span><span class="p">,</span> <span class="s1">&#39;time&#39;</span><span class="p">],</span>
<span class="gp">   ...: </span>         <span class="s1">&#39;sequence&#39;</span><span class="p">:</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span>
<span class="gp">   ...: </span>         <span class="s1">&#39;labels&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;MOTIF&#39;</span><span class="p">,</span> <span class="s1">&#39;SSE03&#39;</span><span class="p">,</span> <span class="s1">&#39;SSE05&#39;</span><span class="p">]}</span>
<span class="gp">   ...: </span><span class="n">df</span> <span class="o">=</span> <span class="n">rs</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">parse_rosetta_file</span><span class="p">(</span><span class="s1">&#39;../rstoolbox/tests/data/input_ssebig.minisilent.gz&#39;</span><span class="p">,</span> <span class="n">rules</span><span class="p">)</span>
<span class="gp">   ...: </span><span class="n">df</span><span class="o">.</span><span class="n">add_reference_sequence</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">baseline</span><span class="o">.</span><span class="n">get_sequence</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">))</span>
<span class="gp">   ...: </span><span class="n">df</span><span class="o">.</span><span class="n">add_reference_shift</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="mi">32</span><span class="p">)</span>
<span class="gp">   ...: </span><span class="n">df</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">   ...: </span>
<span class="gh">Out[2]: </span><span class="go"></span>
<span class="go">    score  ALIGNRMSD  BUNS  COMPRRMSD  C_ni_mtcontacts  C_ni_rmsd  C_ni_trials  MOTIFRMSD  cav_vol  driftRMSD  finalRMSD  packstat C_ni_rmsd_type                            description lbl_MOTIF lbl_SSE03 lbl_SSE05                                                  sequence_C</span>
<span class="go">0 -64.070  0.608      12.0  7.585      4.0              3.301      1.0          0.957      66.602   0.083      3.323      0.544     no_motif       nubinitio_wauto_18326_2pw9C_0001_0001  [C]       [C]       [C]       TTWIKFFAGGTLVEEFEYSSVNWEEIEKRAWKKLGRWKKAEEGDLMIVYPDGKVVSWA</span>
<span class="go">1 -70.981  0.639      12.0  2.410      8.0              1.423      1.0          0.737      0.000    0.094      1.395      0.552     no_motif       nubinitio_wauto_18326_2pw9C_0002_0001  [C]       [C]       [C]       NTWSTNILNGHPKITLLVEERGAEEIHLEWLKKQGLRKKAEENVYTTKLPNGAVKVYG</span>
<span class="go">2 -43.863  0.480      8.0   4.279      6.0              2.110      1.0          0.819      93.641   0.110      2.106      0.575     no_motif       nubinitio_wauto_18326_2pw9C_0003_0001  [C]       [C]       [C]       PRWFIAMGDGVIWEIVLGSEQNLEEIAKKGLKRRGLYKKAEESIYTIIYPDGIAHTFG</span>
</pre></div>
</div>
</div>
<div class="section" id="minimal-mutant">
<h2>Minimal Mutant<a class="headerlink" href="#minimal-mutant" title="Permalink to this headline">¶</a></h2>
<p>We’ve seen multiple ways to identify and view mutations in <a class="reference internal" href="sequence_analysis.html#sequence-analysis"><span class="std std-ref">Sequence Analysis</span></a>. Let’s imagine that we have identified a good decoy candidate but
we want to try all the putative back mutations available. Basically, we ask if a less mutated decoy will perform as well as the one found.</p>
<p>For this, we will take the best scored decoy and we will try to <code class="xref py py-func docutils literal notranslate"><span class="pre">generate_wt_reversions()</span></code> on the residues belonging to strand 3 (<code class="docutils literal notranslate"><span class="pre">SSE03</span></code>) and 5 (<code class="docutils literal notranslate"><span class="pre">SSE05</span></code>):</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [3]: </span><span class="n">kres</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">get_label</span><span class="p">(</span><span class="s1">&#39;SSE03&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">df</span><span class="o">.</span><span class="n">get_label</span><span class="p">(</span><span class="s1">&#39;SSE05&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">   ...: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;score&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">generate_wt_reversions</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">kres</span><span class="p">)</span>
<span class="gp">   ...: </span><span class="n">ds</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">   ...: </span>
<span class="gh">Out[3]: </span><span class="go">16384</span>

<span class="gp">In [4]: </span><span class="n">ds</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
<span class="go">              Out[4]: </span>
<span class="go">                                                   sequence_C                                  description                                                                                                                                                                                                                 mutants_C                                                                                                              mutant_positions_C  mutant_count_C</span>
<span class="go">0  SFWVIIRFRGVIVTIIDIDEPGELERAKEMLLRKGYLKKAEEGETAFFLPKGIAVISA  nubinitio_wauto_18326_2pw9C_0213_0001        E1S,T2F,P3W,Y4V,A5I,A7R,L8F,N9R,D10G,R11V,V12I,I13V,G14T,S15I,S16I,M17D,V18I,L19D,P20E,V21P,D22G,L23E,E24L,F26R,G27A,A28K,G29E,F30M,F32L,G33R,Q34K,I37L,I43G,R44E,E45T,I46A,L47F,V48F,C49L,Q51K,R53I,I54A,S55V,V56I,Y57S  1,2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,26,27,28,29,30,32,33,34,37,43,44,45,46,47,48,49,51,53,54,55,56,57  45            </span>
<span class="go">1  SFWVIIRFRGVIVTIIDIDEPGLEEFGAGFLFGKGYLKKAEEGETAFFLPKGRISVYA  nubinitio_wauto_18326_2pw9C_0213_0001_v0001  E1S,T2F,P3W,Y4V,A5I,A7R,L8F,N9R,D10G,R11V,V12I,I13V,G14T,S15I,S16I,M17D,V18I,L19D,P20E,V21P,D22G,Q34K,I37L,I43G,R44E,E45T,I46A,L47F,V48F,C49L,Q51K                                                                        1,2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,34,37,43,44,45,46,47,48,49,51                                            31            </span>
<span class="go">2  SFWVIIRFRGVIVTIIDIDEPGLEEFGAGFLFGKGYLKKAEEGETAFFLPKGRISVSA  nubinitio_wauto_18326_2pw9C_0213_0001_v0002  E1S,T2F,P3W,Y4V,A5I,A7R,L8F,N9R,D10G,R11V,V12I,I13V,G14T,S15I,S16I,M17D,V18I,L19D,P20E,V21P,D22G,Q34K,I37L,I43G,R44E,E45T,I46A,L47F,V48F,C49L,Q51K,Y57S                                                                   1,2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,34,37,43,44,45,46,47,48,49,51,57                                         32            </span>
<span class="go">3  SFWVIIRFRGVIVTIIDIDEPGLEEFGAGFLFGKGYLKKAEEGETAFFLPKGRISIYA  nubinitio_wauto_18326_2pw9C_0213_0001_v0003  E1S,T2F,P3W,Y4V,A5I,A7R,L8F,N9R,D10G,R11V,V12I,I13V,G14T,S15I,S16I,M17D,V18I,L19D,P20E,V21P,D22G,Q34K,I37L,I43G,R44E,E45T,I46A,L47F,V48F,C49L,Q51K,V56I                                                                   1,2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,34,37,43,44,45,46,47,48,49,51,56                                         32            </span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">To avoid confusion, all scores are unattached from the sequences, as they will need to be recalculated.</p>
</div>
<p>Mind that this will create all the combinatorial options from the selected mutant to the original wild type for the selected residues.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [5]: </span><span class="n">_</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">distplot</span><span class="p">(</span><span class="n">ds</span><span class="p">[</span><span class="s1">&#39;mutant_count_C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>

<span class="gp">In [6]: </span><span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../_images/tutorial_muts_plt1.png"><img alt="../_images/tutorial_muts_plt1.png" src="../_images/tutorial_muts_plt1.png" style="width: 5in;" /></a>
</div>
<div class="section" id="defining-new-mutations-of-interest">
<h2>Defining New Mutations of Interest<a class="headerlink" href="#defining-new-mutations-of-interest" title="Permalink to this headline">¶</a></h2>
<p>Let’s say that, after inspecting scores and visualising structures, there are some key positions for which seems relevant to try some particular residue types, that can also be controlled (also,
remember form <a class="reference internal" href="sequence_analysis.html#sequence-analysis"><span class="std std-ref">Sequence Analysis</span></a> that we can select already decoys with some particular mutations with <a class="reference internal" href="../generated/DesignFrame/rstoolbox.components.DesignFrame.get_sequence_with.html#rstoolbox.components.DesignFrame.get_sequence_with" title="rstoolbox.components.DesignFrame.get_sequence_with"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DesignFrame.get_sequence_with()</span></code></a>). This functionality comes by the
hand of <a class="reference internal" href="../generated/DesignFrame/rstoolbox.components.DesignFrame.generate_mutant_variants.html#rstoolbox.components.DesignFrame.generate_mutant_variants" title="rstoolbox.components.DesignFrame.generate_mutant_variants"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DesignFrame.generate_mutant_variants()</span></code></a>. Let’s now try it for the two best scored decoys.</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [7]: </span><span class="n">mutants</span> <span class="o">=</span> <span class="p">[(</span><span class="mi">20</span><span class="p">,</span> <span class="s2">&quot;AIV&quot;</span><span class="p">),</span> <span class="p">(</span><span class="mi">31</span><span class="p">,</span> <span class="s2">&quot;EDQR&quot;</span><span class="p">)]</span>
<span class="gp">   ...: </span><span class="n">ds</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;score&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">generate_mutant_variants</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">mutants</span><span class="p">)</span>
<span class="gp">   ...: </span><span class="n">ds</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">   ...: </span>
<span class="gh">Out[7]: </span><span class="go">26</span>

<span class="gp">In [8]: </span><span class="n">ds</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
<span class="go">           Out[8]: </span>
<span class="go">                                                   sequence_C                                  description                                                                                                                                                                                                                      mutants_C                                                                                                                 mutant_positions_C  mutant_count_C</span>
<span class="go">0  SFWVIIRFRGVIVTIIDIDEPGELERAKEMLLRKGYLKKAEEGETAFFLPKGIAVISA  nubinitio_wauto_18326_2pw9C_0213_0001        E1S,T2F,P3W,Y4V,A5I,A7R,L8F,N9R,D10G,R11V,V12I,I13V,G14T,S15I,S16I,M17D,V18I,L19D,P20E,V21P,D22G,L23E,E24L,F26R,G27A,A28K,G29E,F30M,F32L,G33R,Q34K,I37L,I43G,R44E,E45T,I46A,L47F,V48F,C49L,Q51K,R53I,I54A,S55V,V56I,Y57S       1,2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,26,27,28,29,30,32,33,34,37,43,44,45,46,47,48,49,51,53,54,55,56,57     45            </span>
<span class="go">1  SFWVIIRFRGVIVTIIDIDAPGELERAKEMELRKGYLKKAEEGETAFFLPKGIAVISA  nubinitio_wauto_18326_2pw9C_0213_0001_v0001  E1S,T2F,P3W,Y4V,A5I,A7R,L8F,N9R,D10G,R11V,V12I,I13V,G14T,S15I,S16I,M17D,V18I,L19D,P20A,V21P,D22G,L23E,E24L,F26R,G27A,A28K,G29E,F30M,L31E,F32L,G33R,Q34K,I37L,I43G,R44E,E45T,I46A,L47F,V48F,C49L,Q51K,R53I,I54A,S55V,V56I,Y57S  1,2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,26,27,28,29,30,31,32,33,34,37,43,44,45,46,47,48,49,51,53,54,55,56,57  46            </span>
<span class="go">2  SFWVIIRFRGVIVTIIDIDAPGELERAKEMDLRKGYLKKAEEGETAFFLPKGIAVISA  nubinitio_wauto_18326_2pw9C_0213_0001_v0002  E1S,T2F,P3W,Y4V,A5I,A7R,L8F,N9R,D10G,R11V,V12I,I13V,G14T,S15I,S16I,M17D,V18I,L19D,P20A,V21P,D22G,L23E,E24L,F26R,G27A,A28K,G29E,F30M,L31D,F32L,G33R,Q34K,I37L,I43G,R44E,E45T,I46A,L47F,V48F,C49L,Q51K,R53I,I54A,S55V,V56I,Y57S  1,2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,26,27,28,29,30,31,32,33,34,37,43,44,45,46,47,48,49,51,53,54,55,56,57  46            </span>
<span class="go">3  SFWVIIRFRGVIVTIIDIDAPGELERAKEMQLRKGYLKKAEEGETAFFLPKGIAVISA  nubinitio_wauto_18326_2pw9C_0213_0001_v0003  E1S,T2F,P3W,Y4V,A5I,A7R,L8F,N9R,D10G,R11V,V12I,I13V,G14T,S15I,S16I,M17D,V18I,L19D,P20A,V21P,D22G,L23E,E24L,F26R,G27A,A28K,G29E,F30M,L31Q,F32L,G33R,Q34K,I37L,I43G,R44E,E45T,I46A,L47F,V48F,C49L,Q51K,R53I,I54A,S55V,V56I,Y57S  1,2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,26,27,28,29,30,31,32,33,34,37,43,44,45,46,47,48,49,51,53,54,55,56,57  46            </span>
</pre></div>
</div>
</div>
<div class="section" id="exhaustive-point-mutant-sampling">
<h2>Exhaustive Point Mutant Sampling<a class="headerlink" href="#exhaustive-point-mutant-sampling" title="Permalink to this headline">¶</a></h2>
<p>An interesting approach could also be to create the appropriate data to exhaustively explore all possible point mutant in all possible
positions. This is extremely easy to do with the <code class="docutils literal notranslate"><span class="pre">rstoolbox</span></code>, generating a total of <code class="docutils literal notranslate"><span class="pre">(len(sequence)</span> <span class="pre">*</span> <span class="pre">19)</span> <span class="pre">+</span> <span class="pre">1</span></code> variants:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [9]: </span><span class="n">point_mutants</span> <span class="o">=</span> <span class="p">[]</span>
<span class="gp">   ...: </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">baseline</span><span class="o">.</span><span class="n">get_reference_sequence</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">))</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
<span class="gp">   ...: </span>    <span class="n">mutation</span> <span class="o">=</span> <span class="p">[(</span><span class="n">i</span><span class="p">,</span> <span class="s2">&quot;*&quot;</span><span class="p">)]</span>
<span class="gp">   ...: </span>    <span class="n">point_mutants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">baseline</span><span class="o">.</span><span class="n">generate_mutant_variants</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">mutation</span><span class="p">))</span>
<span class="gp">   ...: </span><span class="n">point_mutants</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">point_mutants</span><span class="p">)</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="s1">&#39;sequence_C&#39;</span><span class="p">)</span>
<span class="gp">   ...: </span><span class="k">print</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">point_mutants</span><span class="p">[</span><span class="s1">&#39;mutants_C&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">frac</span><span class="o">=</span><span class="mf">0.035</span><span class="p">)))</span>  <span class="c1"># show only some...</span>
<span class="gp">   ...: </span>
<span class="go">A58Y,G14E,V56L,A5I,V18L,R11V,M17A,E1Y,L8T,P3S,V21G,I6N,A5H,S16Y,E41R,E41W,A28H,D10R,M17W,L31R,R44L,N9I,L31W,E41M,L47A,A40P,F30Q,F26D,N9P,A40M,R11W,I13Y,G33S,E1Q,F26M,I6L,V21K,N9C,A58N</span>
</pre></div>
</div>
<p>With small variations, this same loop can be used to generate this exhaustive scanning over particular residues or to only scan
specific residue types (for example to perform an Alanine Scanning).</p>
</div>
<div class="section" id="learning-from-amino-acids-frequency">
<h2>Learning from Amino Acids Frequency<a class="headerlink" href="#learning-from-amino-acids-frequency" title="Permalink to this headline">¶</a></h2>
<p>Instead of manually checking the positions and residue types of interest, one could learn from different sources to improve the type of sequences that one could get.
This new decoy sequences will be scored according to the frequency matrix they were generated from, but any decoy can be mapped to a sequence matrix with
<a class="reference internal" href="../generated/DesignFrame/rstoolbox.components.DesignFrame.score_by_pssm.html#rstoolbox.components.DesignFrame.score_by_pssm" title="rstoolbox.components.DesignFrame.score_by_pssm"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DesignFrame.score_by_pssm()</span></code></a>.</p>
<p>For instance, one could get the information content of the <code class="docutils literal notranslate"><span class="pre">15</span> <span class="pre">top</span> <span class="pre">scored</span></code> decoys and try to obtain a certain amount of mutants (10) from the <code class="docutils literal notranslate"><span class="pre">two</span> <span class="pre">best</span> <span class="pre">packed</span></code> decoy that would
actually follow the statistical rules of that set with <a class="reference internal" href="../generated/DesignFrame/rstoolbox.components.DesignFrame.generate_mutants_from_matrix.html#rstoolbox.components.DesignFrame.generate_mutants_from_matrix" title="rstoolbox.components.DesignFrame.generate_mutants_from_matrix"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DesignFrame.generate_mutants_from_matrix()</span></code></a>:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [10]: </span><span class="n">df15</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;score&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">15</span><span class="p">]</span><span class="o">.</span><span class="n">sequence_frequencies</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">)</span>
<span class="gp">   ....: </span><span class="n">df</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;packstat&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="bp">False</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">generate_mutants_from_matrix</span><span class="p">(</span><span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="n">df15</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="gp">   ....: </span>
<span class="gh">Out[10]: </span><span class="go"></span>
<span class="go">[                                   description                                                  sequence_C  pssm_score_C</span>
<span class="go"> 0  nubinitio_wauto_18326_2pw9C_0260_0001_v0001  RTPVEVKFNGELVTMLDSNGRRQKEIAYAELLKQGLWKKAEEGKYTIIYPDGWVVFVG  23.200000   </span>
<span class="go"> 1  nubinitio_wauto_18326_2pw9C_0260_0001_v0002  PTVVVVRLGGQPISRQLRDLINLEEIAKEFMRKLGYAKKAEEGQKVVIEPDGSIISAD  24.866667   </span>
<span class="go"> 2  nubinitio_wauto_18326_2pw9C_0260_0001_v0003  KKWLVVMLRGVPKIVIYVLNSNLEEEAREYLREQGLWKKAEEGTYTIVYSKGYIRTSA  23.933333   </span>
<span class="go"> 3  nubinitio_wauto_18326_2pw9C_0260_0001_v0004  GQWIYVMFRGVPIKVMDIKDKNEEEEAWEMLKKKGVWKKAEEGKYVYILPNGYVITAA  24.733333   </span>
<span class="go"> 4  nubinitio_wauto_18326_2pw9C_0260_0001_v0005  LFWVEFFFNGVPTFMLLSEDIGEEEIARRIFKKKGLLKKAEEGTYVMYLPDGIVEISD  25.666667   </span>
<span class="go"> 5  nubinitio_wauto_18326_2pw9C_0260_0001_v0006  NTWVLVRFRGVPKTVLLTNNRGQLEIAREEMRKLGLMKKAEEDIYIRFLPNGIVIISP  24.400000   </span>
<span class="go"> 6  nubinitio_wauto_18326_2pw9C_0260_0001_v0007  PFWVVVELNGVPIIVQIVVPIGEEEIAREYQKKQGVWKKAEEGQEGEYYPDGVLLIYG  25.333333   </span>
<span class="go"> 7  nubinitio_wauto_18326_2pw9C_0260_0001_v0008  PQVFVFETGGEPKTIQIINDKNQEELAEKEYKRLGQGKKAEEGKTVMIYPNGVIYFYG  23.266667   </span>
<span class="go"> 8  nubinitio_wauto_18326_2pw9C_0260_0001_v0009  PTWVEFFFNGEPIMILLTTGSGLKERAWEYLKKMGQLKKAEEGEYVVFSPNGWIEHYA  24.866667   </span>
<span class="go"> 9  nubinitio_wauto_18326_2pw9C_0260_0001_v0010  PQWLVFEVNGVLKTMQLRVNPNLLEIAYRLLRKLGLWKKAEEGTETIILSDGIVLSSE  24.000000   ,</span>
<span class="go">                                    description                                                  sequence_C  pssm_score_C</span>
<span class="go"> 0  nubinitio_wauto_18326_2pw9C_0278_0001_v0001  KKVFEVRYNGVPTSVLLTNNTGLEEIALRLLKRQGQWKKAEEGEFSIIYPNGAVIFYP  24.666667   </span>
<span class="go"> 1  nubinitio_wauto_18326_2pw9C_0278_0001_v0002  TTLIKFMLNGVPVTMQLSEGSNLLEYAYKLQKKQGLLKKAEEGKTVYIYSKGEVRFWA  23.733333   </span>
<span class="go"> 2  nubinitio_wauto_18326_2pw9C_0278_0001_v0003  PTWVVFFTNGVPKSRILTEDSNLEEIGEREMKKLGVWKKAEEGKYVIVKPDGIVIIVG  26.466667   </span>
<span class="go"> 3  nubinitio_wauto_18326_2pw9C_0278_0001_v0004  KTWVEVRFNGEPITISLLNGRNIEEAAYRELKKKGLLKKAEEGKEVEVYPSGIILHAA  26.866667   </span>
<span class="go"> 4  nubinitio_wauto_18326_2pw9C_0278_0001_v0005  GIWVVIRVNGEPISVLATKGKNLEEIAREEFRKLGLMKKAEENKYVEEYPKGVIWIYD  25.533333   </span>
<span class="go"> 5  nubinitio_wauto_18326_2pw9C_0278_0001_v0006  TFWVEVKLNGVPKIVMLVENINEKEIAREELKKMGLLKKAEESKYVRIYPDGVVYTFG  26.866667   </span>
<span class="go"> 6  nubinitio_wauto_18326_2pw9C_0278_0001_v0007  LTPFVIRVNGQPITMILVVNSNEEERAREELKRKGLLKKAEEGKFVYFEPKGVAYSYG  25.466667   </span>
<span class="go"> 7  nubinitio_wauto_18326_2pw9C_0278_0001_v0008  STWVYFRVNGEPISVVLVNNRNQEERARREQRRKGLWKKAEESEEVIIYSDGVVYSTG  24.933333   </span>
<span class="go"> 8  nubinitio_wauto_18326_2pw9C_0278_0001_v0009  SQWVVVRVRGVPIKRILVTPKNQEEIAYEELKKQGKWKKAEEGTTAIILPKGEIWFYD  25.933333   </span>
<span class="go"> 9  nubinitio_wauto_18326_2pw9C_0278_0001_v0010  PIPVEVKFNGVPIKMIAIVDIKEEKYAREFLKKQGLAKKAEEDKTVIFLPDGELWSYA  23.866667   ]</span>
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">While the generation of <strong>the best scored sequence</strong> according to a matrix is quite straight forward, the generation of the N best is not. Thus,
the way a frequency matrix is applied in this scenario is that the frequency corrects a randomised selection per position. Statistically, this will make
most sequences obtained on the <em>upper side</em> of the matrix score.</p>
</div>
</div>
<div class="section" id="creating-the-new-mutants">
<h2>Creating the New Mutants<a class="headerlink" href="#creating-the-new-mutants" title="Permalink to this headline">¶</a></h2>
<p>After generating new variations, those can be run in <strong>Rosetta</strong> to obtain their corresponding scores with <code class="xref py py-class docutils literal notranslate"><span class="pre">DesignFrame.apply_resfile</span></code>. By default, this function will run
<em>fixed backbone</em> design with a script such as:</p>
<div class="highlight-ipython notranslate"><div class="highlight"><pre><span></span><span class="gp">In [11]: </span><span class="k">print</span><span class="p">(</span><span class="n">rs</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">mutations</span><span class="p">())</span>
<span class="go">&lt;ROSETTASCRIPTS&gt;</span>
<span class="go">    &lt;TASKOPERATIONS&gt;</span>
<span class="go">        &lt;ReadResfile name=&quot;targets&quot; filename=&quot;%%resfile%%&quot;/&gt;</span>
<span class="go">    &lt;/TASKOPERATIONS&gt;</span>
<span class="go">    &lt;MOVERS&gt;</span>
<span class="go">        &lt;PackRotamersMover name=&quot;packrot&quot; task_operations=&quot;targets&quot; /&gt;</span>
<span class="go">        &lt;AddJobPairData name=&quot;annotate&quot; value_type=&quot;string&quot;</span>
<span class="go">                        key=&quot;resfile_A&quot; value=&quot;%%resfile%%&quot; /&gt;</span>
<span class="go">    &lt;/MOVERS&gt;</span>
<span class="go">    &lt;PROTOCOLS&gt;</span>
<span class="go">        &lt;Add mover=&quot;packrot&quot; /&gt;</span>
<span class="go">        &lt;Add mover=&quot;annotate&quot; /&gt;</span>
<span class="go">    &lt;/PROTOCOLS&gt;</span>
<span class="go">&lt;/ROSETTASCRIPTS&gt;</span>
</pre></div>
</div>
<p>But the user can provide more complex <strong>RosettaScripts</strong> to execute, as long as they follow the same restrictions as this one:</p>
<ol class="arabic simple">
<li>They contain the <a class="reference external" href="https://www.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/Movers/movers_pages/AddJobPairDataMover">AddJobPairData Mover</a>.</li>
<li>They target the <strong>resfile</strong> with the <code class="docutils literal notranslate"><span class="pre">script_var</span></code> <code class="docutils literal notranslate"><span class="pre">%%resfile%%</span></code>.</li>
</ol>
</div>
</div>


    </div>
      
  </div>
</div>

<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
        <br/>
        
      
    </p>
    <p>
        &copy; Copyright 2018, Jaume Bonet.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.2.<br/>
    </p>
  </div>
</footer>
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-74877325-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-74877325-2');
</script>

  </body>
</html>