<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>API reference &#8212; rstoolbox 0.9.3 documentation</title>
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/style.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/functions.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="_static/bootstrap-3.3.7/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="rstoolbox.components.Selection" href="generated/rstoolbox.components.Selection.html" />
    <link rel="prev" title="RosettaScripts" href="rosettascripts.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head><body>

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          rstoolbox</a>
        <span class="navbar-text navbar-version pull-left"><b>0.9.3</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="scripts.html">Scripts</a></li>
                <li><a href="tutorial.html">Tutorial</a></li>
                <li><a href="#">API</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">An introduction to rstoolbox</a></li>
<li class="toctree-l1"><a class="reference internal" href="highlights.html">Highlighted works</a></li>
<li class="toctree-l1"><a class="reference internal" href="installing.html">Installing and Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="scripts.html">Available direct executables</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="tutorial.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="rosettascripts.html">RosettaScripts</a></li>
<li class="toctree-l1"><a class="reference internal" href="whatsnew.html">What’s new</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">API reference</a><ul>
<li><a class="reference internal" href="#components">Components</a></li>
<li><a class="reference internal" href="#io-sequence">IO: Sequence</a></li>
<li><a class="reference internal" href="#io-rosetta">IO: Rosetta</a></li>
<li><a class="reference internal" href="#io-experiments">IO: Experiments</a></li>
<li><a class="reference internal" href="#analysis">Analysis</a></li>
<li><a class="reference internal" href="#plot">Plot</a></li>
<li><a class="reference internal" href="#plot-experiments">Plot: Experiments</a></li>
<li><a class="reference internal" href="#utils-plot">Utils: Plot</a></li>
<li><a class="reference internal" href="#utils-contextualize">Utils: Contextualize</a></li>
<li><a class="reference internal" href="#utils-transforms">Utils: Transforms</a></li>
<li><a class="reference internal" href="#utils-rosettascript">Utils: RosettaScript</a></li>
<li><a class="reference internal" href="#utils-experiments">Utils: Experiments</a></li>
<li><a class="reference internal" href="#internals-functions-for-developers">Internals: Functions for Developers</a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  <div class="section" id="api-reference">
<span id="api-ref"></span><h1>API reference<a class="headerlink" href="#api-reference" title="Permalink to this headline">¶</a></h1>
<div class="section" id="components">
<span id="grid-api"></span><h2>Components<a class="headerlink" href="#components" title="Permalink to this headline">¶</a></h2>
<p>These are the list of dedicated objects provided to manage design data. They can be called through <code class="docutils literal notranslate"><span class="pre">rstoolbox.components</span></code>.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.components.Selection.html#rstoolbox.components.Selection" title="rstoolbox.components.Selection"><code class="xref py py-obj docutils literal notranslate"><span class="pre">Selection</span></code></a>([selection])</td>
<td>Complex management of residue selection from a sequence.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.components.SelectionContainer.html#rstoolbox.components.SelectionContainer" title="rstoolbox.components.SelectionContainer"><code class="xref py py-obj docutils literal notranslate"><span class="pre">SelectionContainer</span></code></a>(*args)</td>
<td>Helper class to manage representation of selectors in <a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/index.html#module-pandas" title="(in pandas v0.24.2)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">pandas</span></code></a>.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.components.DesignSeries.html#rstoolbox.components.DesignSeries" title="rstoolbox.components.DesignSeries"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DesignSeries</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>The <a class="reference internal" href="generated/rstoolbox.components.DesignSeries.html#rstoolbox.components.DesignSeries" title="rstoolbox.components.DesignSeries"><code class="xref py py-class docutils literal notranslate"><span class="pre">DesignSeries</span></code></a> extends the <a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html#pandas.Series" title="(in pandas v0.24.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Series</span></code></a> adding some functionalities in order to improve its usability in the analysis of a single design decoys.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.components.DesignFrame.html#rstoolbox.components.DesignFrame" title="rstoolbox.components.DesignFrame"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DesignFrame</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>The <a class="reference internal" href="generated/rstoolbox.components.DesignFrame.html#rstoolbox.components.DesignFrame" title="rstoolbox.components.DesignFrame"><code class="xref py py-class docutils literal notranslate"><span class="pre">DesignFrame</span></code></a> extends the <a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v0.24.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></a> adding some functionalities in order to improve its usability in the analysis of sets of design decoys.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.components.SequenceFrame.html#rstoolbox.components.SequenceFrame" title="rstoolbox.components.SequenceFrame"><code class="xref py py-obj docutils literal notranslate"><span class="pre">SequenceFrame</span></code></a>(*args,&nbsp;**kwargs)</td>
<td>Per position frequency occurrence for a set of decoys.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.components.FragmentFrame.html#rstoolbox.components.FragmentFrame" title="rstoolbox.components.FragmentFrame"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FragmentFrame</span></code></a>(*args,&nbsp;**kw)</td>
<td>Data container for Fragment data.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="io-sequence">
<h2>IO: Sequence<a class="headerlink" href="#io-sequence" title="Permalink to this headline">¶</a></h2>
<p>Helper functions to read/write direct sequence information. They can be called through <code class="docutils literal notranslate"><span class="pre">rstoolbox.io</span></code>.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.io.read_fasta.html#rstoolbox.io.read_fasta" title="rstoolbox.io.read_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_fasta</span></code></a>(filename[,&nbsp;expand,&nbsp;multi,&nbsp;defchain])</td>
<td>Reads one or more <strong>FASTA</strong> files and returns the appropiate object containing the requested data: the <a class="reference internal" href="generated/rstoolbox.components.DesignFrame.html#rstoolbox.components.DesignFrame" title="rstoolbox.components.DesignFrame"><code class="xref py py-class docutils literal notranslate"><span class="pre">DesignFrame</span></code></a>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.io.write_fasta.html#rstoolbox.io.write_fasta" title="rstoolbox.io.write_fasta"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_fasta</span></code></a>(df,&nbsp;seqID[,&nbsp;separator,&nbsp;…])</td>
<td>Writes fasta files of the selected decoys.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.io.write_clustalw.html#rstoolbox.io.write_clustalw" title="rstoolbox.io.write_clustalw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_clustalw</span></code></a>(df,&nbsp;seqID[,&nbsp;filename])</td>
<td>Write sequences of selected designs as a CLUSTALW alignment.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.io.write_mutant_alignments.html#rstoolbox.io.write_mutant_alignments" title="rstoolbox.io.write_mutant_alignments"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_mutant_alignments</span></code></a>(df,&nbsp;seqID[,&nbsp;filename])</td>
<td>Writes a text file containing only the positions changed with respect to the <code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code>.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.io.read_hmmsearch.html#rstoolbox.io.read_hmmsearch" title="rstoolbox.io.read_hmmsearch"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_hmmsearch</span></code></a>(filename)</td>
<td>Read output from <code class="docutils literal notranslate"><span class="pre">hmmsearch</span></code> or <code class="docutils literal notranslate"><span class="pre">hmmscan</span></code>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.io.pymol_mutant_selector.html#rstoolbox.io.pymol_mutant_selector" title="rstoolbox.io.pymol_mutant_selector"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pymol_mutant_selector</span></code></a>(df)</td>
<td>Generate selectors for the mutations in target decoys.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="io-rosetta">
<h2>IO: Rosetta<a class="headerlink" href="#io-rosetta" title="Permalink to this headline">¶</a></h2>
<p>Helper functions to read/write data generated with <a class="reference external" href="https://www.rosettacommons.org/">Rosetta</a>. They can be called through <code class="docutils literal notranslate"><span class="pre">rstoolbox.io</span></code>.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.io.parse_rosetta_file.html#rstoolbox.io.parse_rosetta_file" title="rstoolbox.io.parse_rosetta_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_rosetta_file</span></code></a>(filename[,&nbsp;description,&nbsp;…])</td>
<td>Read a Rosetta score or silent file and returns the design population in a <a class="reference internal" href="generated/rstoolbox.components.DesignFrame.html#rstoolbox.components.DesignFrame" title="rstoolbox.components.DesignFrame"><code class="xref py py-class docutils literal notranslate"><span class="pre">DesignFrame</span></code></a>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.io.parse_rosetta_json.html#rstoolbox.io.parse_rosetta_json" title="rstoolbox.io.parse_rosetta_json"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_rosetta_json</span></code></a>(filename)</td>
<td>Read a json formated rosetta score file.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.io.parse_rosetta_contacts.html#rstoolbox.io.parse_rosetta_contacts" title="rstoolbox.io.parse_rosetta_contacts"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_rosetta_contacts</span></code></a>(filename)</td>
<td>Read a residue contact file as generated by <strong>ContactMapMover</strong>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.io.parse_rosetta_fragments.html#rstoolbox.io.parse_rosetta_fragments" title="rstoolbox.io.parse_rosetta_fragments"><code class="xref py py-obj docutils literal notranslate"><span class="pre">parse_rosetta_fragments</span></code></a>(filename[,&nbsp;source])</td>
<td>Read a Rosetta fragment-file and return the appropiate <a class="reference internal" href="generated/rstoolbox.components.FragmentFrame.html#rstoolbox.components.FragmentFrame" title="rstoolbox.components.FragmentFrame"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentFrame</span></code></a>.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.io.write_rosetta_fragments.html#rstoolbox.io.write_rosetta_fragments" title="rstoolbox.io.write_rosetta_fragments"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_rosetta_fragments</span></code></a>(df[,&nbsp;frag_size,&nbsp;…])</td>
<td>Writes a Rosetta fragment-file (new format) from an appropiate <a class="reference internal" href="generated/rstoolbox.components.FragmentFrame.html#rstoolbox.components.FragmentFrame" title="rstoolbox.components.FragmentFrame"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentFrame</span></code></a>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.io.write_fragment_sequence_profiles.html#rstoolbox.io.write_fragment_sequence_profiles" title="rstoolbox.io.write_fragment_sequence_profiles"><code class="xref py py-obj docutils literal notranslate"><span class="pre">write_fragment_sequence_profiles</span></code></a>(df[,&nbsp;…])</td>
<td>Write a sequence profile from <a class="reference internal" href="generated/rstoolbox.components.FragmentFrame.html#rstoolbox.components.FragmentFrame" title="rstoolbox.components.FragmentFrame"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentFrame</span></code></a> to load into Rosetta’s <strong>SeqprofConsensus</strong>.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.io.get_sequence_and_structure.html#rstoolbox.io.get_sequence_and_structure" title="rstoolbox.io.get_sequence_and_structure"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequence_and_structure</span></code></a>(pdbfile[,&nbsp;…])</td>
<td>Provided a PDB file, it will run a small <strong>RosettaScript</strong> to capture its sequence and structure, i.e.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.io.make_structures.html#rstoolbox.io.make_structures" title="rstoolbox.io.make_structures"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_structures</span></code></a>(df[,&nbsp;outdir,&nbsp;tagsfilename,&nbsp;…])</td>
<td>Extract the selected decoys (if any).</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="io-experiments">
<h2>IO: Experiments<a class="headerlink" href="#io-experiments" title="Permalink to this headline">¶</a></h2>
<p>Helper functions to read/write data generated through wedlab experiments. They can be called through <code class="docutils literal notranslate"><span class="pre">rstoolbox.io</span></code>.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.io.read_SPR.html#rstoolbox.io.read_SPR" title="rstoolbox.io.read_SPR"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_SPR</span></code></a>(filename)</td>
<td>Reads <strong>Surface Plasmon Resonance</strong> data.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.io.read_CD.html#rstoolbox.io.read_CD" title="rstoolbox.io.read_CD"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_CD</span></code></a>(dirname[,&nbsp;prefix,&nbsp;invert_temp,&nbsp;…])</td>
<td>Read <a class="reference external" href="https://www.wikiwand.com/en/Circular_dichroism">Circular Dichroism</a> data for multiple temperatures.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.io.read_MALS.html#rstoolbox.io.read_MALS" title="rstoolbox.io.read_MALS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_MALS</span></code></a>(filename[,&nbsp;mmfile])</td>
<td>Read data from <a class="reference external" href="https://www.wikiwand.com/en/Multiangle_light_scattering">Multi-Angle Light Scattering</a> data.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.io.read_fastq.html#rstoolbox.io.read_fastq" title="rstoolbox.io.read_fastq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">read_fastq</span></code></a>(filename[,&nbsp;seqID])</td>
<td>Reads a FASTQ file and stores the ID together with the sequence.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="analysis">
<h2>Analysis<a class="headerlink" href="#analysis" title="Permalink to this headline">¶</a></h2>
<p>Helper functions for sequence analysis. They can be called through <code class="docutils literal notranslate"><span class="pre">rstoolbox.analysis</span></code>.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.analysis.sequential_frequencies.html#rstoolbox.analysis.sequential_frequencies" title="rstoolbox.analysis.sequential_frequencies"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequential_frequencies</span></code></a>(df,&nbsp;seqID[,&nbsp;query,&nbsp;…])</td>
<td>Generates a <a class="reference internal" href="generated/rstoolbox.components.SequenceFrame.html#rstoolbox.components.SequenceFrame" title="rstoolbox.components.SequenceFrame"><code class="xref py py-class docutils literal notranslate"><span class="pre">SequenceFrame</span></code></a> for the frequencies of the sequences in the <a class="reference internal" href="generated/rstoolbox.components.DesignFrame.html#rstoolbox.components.DesignFrame" title="rstoolbox.components.DesignFrame"><code class="xref py py-class docutils literal notranslate"><span class="pre">DesignFrame</span></code></a> with <code class="docutils literal notranslate"><span class="pre">seqID</span></code> identifier.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.analysis.sequence_similarity.html#rstoolbox.analysis.sequence_similarity" title="rstoolbox.analysis.sequence_similarity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence_similarity</span></code></a>(df,&nbsp;seqID[,&nbsp;…])</td>
<td>Evaluate the sequence similarity between each decoy and the <code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code> for a given <code class="docutils literal notranslate"><span class="pre">seqID</span></code>.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.analysis.positional_sequence_similarity.html#rstoolbox.analysis.positional_sequence_similarity" title="rstoolbox.analysis.positional_sequence_similarity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">positional_sequence_similarity</span></code></a>(df[,&nbsp;seqID,&nbsp;…])</td>
<td>Per position identity and similarity against a <code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.analysis.binary_similarity.html#rstoolbox.analysis.binary_similarity" title="rstoolbox.analysis.binary_similarity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">binary_similarity</span></code></a>(df,&nbsp;seqID[,&nbsp;key_residues,&nbsp;…])</td>
<td>Binary profile for each design sequence against the <code class="docutils literal notranslate"><span class="pre">reference_sequence</span></code>.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.analysis.binary_overlap.html#rstoolbox.analysis.binary_overlap" title="rstoolbox.analysis.binary_overlap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">binary_overlap</span></code></a>(df,&nbsp;seqID[,&nbsp;key_residues,&nbsp;matrix])</td>
<td>Overlap the binary similarity representation of all decoys in a <a class="reference internal" href="generated/rstoolbox.components.DesignFrame.html#rstoolbox.components.DesignFrame" title="rstoolbox.components.DesignFrame"><code class="xref py py-class docutils literal notranslate"><span class="pre">DesignFrame</span></code></a>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.analysis.positional_enrichment.html#rstoolbox.analysis.positional_enrichment" title="rstoolbox.analysis.positional_enrichment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">positional_enrichment</span></code></a>(df,&nbsp;other,&nbsp;seqID)</td>
<td>Calculates per-residue enrichment from sequences in the first <a class="reference internal" href="generated/rstoolbox.components.DesignFrame.html#rstoolbox.components.DesignFrame" title="rstoolbox.components.DesignFrame"><code class="xref py py-class docutils literal notranslate"><span class="pre">DesignFrame</span></code></a> with respect to the second.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.analysis.positional_structural_count.html#rstoolbox.analysis.positional_structural_count" title="rstoolbox.analysis.positional_structural_count"><code class="xref py py-obj docutils literal notranslate"><span class="pre">positional_structural_count</span></code></a>(df[,&nbsp;seqID,&nbsp;…])</td>
<td>Percentage of secondary structure types for each sequence position of all decoys.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.analysis.positional_structural_identity.html#rstoolbox.analysis.positional_structural_identity" title="rstoolbox.analysis.positional_structural_identity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">positional_structural_identity</span></code></a>(df[,&nbsp;seqID,&nbsp;…])</td>
<td>Per position evaluation of how many times the provided data matches the expected <code class="docutils literal notranslate"><span class="pre">reference_structure</span></code>.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.analysis.secondary_structure_percentage.html#rstoolbox.analysis.secondary_structure_percentage" title="rstoolbox.analysis.secondary_structure_percentage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">secondary_structure_percentage</span></code></a>(df,&nbsp;seqID[,&nbsp;…])</td>
<td>Calculate the percentage of the different secondary structure types.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.analysis.selector_percentage.html#rstoolbox.analysis.selector_percentage" title="rstoolbox.analysis.selector_percentage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">selector_percentage</span></code></a>(df,&nbsp;seqID,&nbsp;key_residues)</td>
<td>Calculate the percentage coverage of a <a class="reference internal" href="generated/rstoolbox.components.Selection.html#rstoolbox.components.Selection" title="rstoolbox.components.Selection"><code class="xref py py-class docutils literal notranslate"><span class="pre">Selection</span></code></a> over the sequence.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.analysis.label_percentage.html#rstoolbox.analysis.label_percentage" title="rstoolbox.analysis.label_percentage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">label_percentage</span></code></a>(df,&nbsp;seqID,&nbsp;label)</td>
<td>Calculate the percentage coverage of a <code class="docutils literal notranslate"><span class="pre">label</span></code> over the sequence.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.analysis.cumulative.html#rstoolbox.analysis.cumulative" title="rstoolbox.analysis.cumulative"><code class="xref py py-obj docutils literal notranslate"><span class="pre">cumulative</span></code></a>(values[,&nbsp;bins,&nbsp;max_count,&nbsp;…])</td>
<td>Generates, for a given list of values, its cumulative distribution values.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="plot">
<h2>Plot<a class="headerlink" href="#plot" title="Permalink to this headline">¶</a></h2>
<p>Once the data is loaded in the different <strong>components</strong>, it is ready to use into any
plotting library, but some special plotting alternatives are offered through <code class="docutils literal notranslate"><span class="pre">rstoolbox.plot</span></code>.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.plot.multiple_distributions.html#rstoolbox.plot.multiple_distributions" title="rstoolbox.plot.multiple_distributions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">multiple_distributions</span></code></a>(df,&nbsp;fig,&nbsp;grid[,&nbsp;…])</td>
<td>Automatically plot boxplot distributions for multiple score types of the decoy population.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.plot.sequence_frequency_plot.html#rstoolbox.plot.sequence_frequency_plot" title="rstoolbox.plot.sequence_frequency_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence_frequency_plot</span></code></a>(df,&nbsp;seqID,&nbsp;ax[,&nbsp;…])</td>
<td>Makes a heatmap subplot into the provided axis showing the sequence distribution of each residue type for each position.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.plot.logo_plot.html#rstoolbox.plot.logo_plot" title="rstoolbox.plot.logo_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">logo_plot</span></code></a>(df,&nbsp;seqID[,&nbsp;refseq,&nbsp;key_residues,&nbsp;…])</td>
<td>Generates full figure classic <strong>LOGO</strong> plots.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.plot.logo_plot_in_axis.html#rstoolbox.plot.logo_plot_in_axis" title="rstoolbox.plot.logo_plot_in_axis"><code class="xref py py-obj docutils literal notranslate"><span class="pre">logo_plot_in_axis</span></code></a>(df,&nbsp;seqID,&nbsp;ax[,&nbsp;refseq,&nbsp;…])</td>
<td>Generates classic <strong>LOGO</strong> plot in a given axis.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.plot.positional_sequence_similarity_plot.html#rstoolbox.plot.positional_sequence_similarity_plot" title="rstoolbox.plot.positional_sequence_similarity_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">positional_sequence_similarity_plot</span></code></a>(df,&nbsp;ax)</td>
<td>Generates a plot covering the amount of identities and positives matches from a population of designs to a reference sequence according to a substitution matrix.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.plot.per_residue_matrix_score_plot.html#rstoolbox.plot.per_residue_matrix_score_plot" title="rstoolbox.plot.per_residue_matrix_score_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">per_residue_matrix_score_plot</span></code></a>(df,&nbsp;seqID,&nbsp;ax)</td>
<td>Plot a linear representation of the scoring obtained by applying a substitution matrix.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.plot.positional_structural_similarity_plot.html#rstoolbox.plot.positional_structural_similarity_plot" title="rstoolbox.plot.positional_structural_similarity_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">positional_structural_similarity_plot</span></code></a>(df,&nbsp;ax)</td>
<td>Generates a bar plot for positional prevalence of secondary structure elements.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.plot.plot_fragments.html#rstoolbox.plot.plot_fragments" title="rstoolbox.plot.plot_fragments"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_fragments</span></code></a>(small_frags,&nbsp;large_frags,&nbsp;…)</td>
<td>Plot RMSD quality of a pair of <a class="reference internal" href="generated/rstoolbox.components.FragmentFrame.html#rstoolbox.components.FragmentFrame" title="rstoolbox.components.FragmentFrame"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentFrame</span></code></a> in two provided axis.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.plot.plot_fragment_profiles.html#rstoolbox.plot.plot_fragment_profiles" title="rstoolbox.plot.plot_fragment_profiles"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_fragment_profiles</span></code></a>(fig,&nbsp;small_frags,&nbsp;…)</td>
<td>Plots a full summary of the a <a class="reference internal" href="generated/rstoolbox.components.FragmentFrame.html#rstoolbox.components.FragmentFrame" title="rstoolbox.components.FragmentFrame"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentFrame</span></code></a> quality with sequence and expected secondary structure match.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.plot.plot_alignment.html#rstoolbox.plot.plot_alignment" title="rstoolbox.plot.plot_alignment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_alignment</span></code></a>(df,&nbsp;seqID,&nbsp;ax[,&nbsp;line_break,&nbsp;…])</td>
<td>Make an image representing the alignment of sequences with higlights to mutant positions.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.plot.plot_ramachandran.html#rstoolbox.plot.plot_ramachandran" title="rstoolbox.plot.plot_ramachandran"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_ramachandran</span></code></a>(df,&nbsp;seqID,&nbsp;fig[,&nbsp;grid,&nbsp;…])</td>
<td>Generates a ramachandran plot in RAMPAGE style.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.plot.plot_ramachandran_single.html#rstoolbox.plot.plot_ramachandran_single" title="rstoolbox.plot.plot_ramachandran_single"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_ramachandran_single</span></code></a>(df,&nbsp;seqID,&nbsp;ax[,&nbsp;…])</td>
<td>Plot only one of the 4 ramachandran plots in RAMPAGE format.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.plot.plot_dssp_vs_psipred.html#rstoolbox.plot.plot_dssp_vs_psipred" title="rstoolbox.plot.plot_dssp_vs_psipred"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_dssp_vs_psipred</span></code></a>(df,&nbsp;seqID,&nbsp;ax)</td>
<td>Generates a horizontal heatmap showing differences in psipred predictions to dssp assignments.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="plot-experiments">
<h2>Plot: Experiments<a class="headerlink" href="#plot-experiments" title="Permalink to this headline">¶</a></h2>
<p>Plot data obtained from experimental procedures. Accessible through <code class="docutils literal notranslate"><span class="pre">rstoolbox.plot</span></code>.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.plot.plot_96wells.html#rstoolbox.plot.plot_96wells" title="rstoolbox.plot.plot_96wells"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_96wells</span></code></a>([cdata,&nbsp;sdata,&nbsp;bdata,&nbsp;bcolors,&nbsp;…])</td>
<td>Plot data of a 96 well plate into an equivalent-shaped plot.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.plot.plot_thermal_melt.html#rstoolbox.plot.plot_thermal_melt" title="rstoolbox.plot.plot_thermal_melt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_thermal_melt</span></code></a>(df,&nbsp;ax[,&nbsp;linecolor,&nbsp;…])</td>
<td>Plot <a class="reference external" href="https://www.wikiwand.com/en/Thermal_shift_assay">Thermal Melt</a> data.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.plot.plot_MALS.html#rstoolbox.plot.plot_MALS" title="rstoolbox.plot.plot_MALS"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_MALS</span></code></a>(df,&nbsp;ax[,&nbsp;uvcolor,&nbsp;lscolor,&nbsp;…])</td>
<td>Plot <a class="reference external" href="https://www.wikiwand.com/en/Multiangle_light_scattering">Multi-Angle Light Scattering</a> data.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.plot.plot_CD.html#rstoolbox.plot.plot_CD" title="rstoolbox.plot.plot_CD"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_CD</span></code></a>(df,&nbsp;ax[,&nbsp;color,&nbsp;wavelengths,&nbsp;sample])</td>
<td>Plot <a class="reference external" href="https://www.wikiwand.com/en/Circular_dichroism">Circular Dichroism</a> data.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.plot.plot_SPR.html#rstoolbox.plot.plot_SPR" title="rstoolbox.plot.plot_SPR"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_SPR</span></code></a>(df,&nbsp;ax[,&nbsp;datacolor,&nbsp;fitcolor,&nbsp;…])</td>
<td>Plot <a class="reference external" href="https://www.wikiwand.com/en/Surface_plasmon_resonance">Surface Plasmon Resonance</a> data.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="utils-plot">
<h2>Utils: Plot<a class="headerlink" href="#utils-plot" title="Permalink to this headline">¶</a></h2>
<p>Special functions to help personalise your plot easily can be loaded through <code class="docutils literal notranslate"><span class="pre">rstoolbox.utils</span></code>.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.utils.format_Ipython.html#rstoolbox.utils.format_Ipython" title="rstoolbox.utils.format_Ipython"><code class="xref py py-obj docutils literal notranslate"><span class="pre">format_Ipython</span></code></a>()</td>
<td>Ensure <code class="docutils literal notranslate"><span class="pre">monospace</span></code> representation of <a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v0.24.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></a> in <strong>Jupyter Notebooks</strong>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.utils.use_qgrid.html#rstoolbox.utils.use_qgrid" title="rstoolbox.utils.use_qgrid"><code class="xref py py-obj docutils literal notranslate"><span class="pre">use_qgrid</span></code></a>(df,&nbsp;**kwargs)</td>
<td>Create a <code class="docutils literal notranslate"><span class="pre">QgridWidget</span></code> object from the <a class="reference external" href="https://qgrid.readthedocs.io/en/latest/">qgrid library</a> in <strong>Jupyter Notebooks</strong>.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.utils.add_left_title.html#rstoolbox.utils.add_left_title" title="rstoolbox.utils.add_left_title"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_left_title</span></code></a>(ax,&nbsp;title,&nbsp;**kwargs)</td>
<td>Add a centered title on the left of the selected axis.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.utils.add_right_title.html#rstoolbox.utils.add_right_title" title="rstoolbox.utils.add_right_title"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_right_title</span></code></a>(ax,&nbsp;title,&nbsp;**kwargs)</td>
<td>Add a centered title on rigth of the selected axis.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.utils.add_top_title.html#rstoolbox.utils.add_top_title" title="rstoolbox.utils.add_top_title"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_top_title</span></code></a>(ax,&nbsp;title,&nbsp;**kwargs)</td>
<td>Add a centered title on top of the selected axis.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.utils.edit_legend_text.html#rstoolbox.utils.edit_legend_text" title="rstoolbox.utils.edit_legend_text"><code class="xref py py-obj docutils literal notranslate"><span class="pre">edit_legend_text</span></code></a>(ax,&nbsp;labels[,&nbsp;title])</td>
<td>Change the labels and title of a legend.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.utils.add_white_to_cmap.html#rstoolbox.utils.add_white_to_cmap" title="rstoolbox.utils.add_white_to_cmap"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_white_to_cmap</span></code></a>([color,&nbsp;cmap,&nbsp;n_colors])</td>
<td>Generate a new colormap with white as first instance.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.utils.color_variant.html#rstoolbox.utils.color_variant" title="rstoolbox.utils.color_variant"><code class="xref py py-obj docutils literal notranslate"><span class="pre">color_variant</span></code></a>(color[,&nbsp;brightness_offset])</td>
<td>Make a color darker or lighter.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="utils-contextualize">
<h2>Utils: Contextualize<a class="headerlink" href="#utils-contextualize" title="Permalink to this headline">¶</a></h2>
<p>Functions aimed to help assess a design population in the context of known protein structures.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.utils.load_refdata.html#rstoolbox.utils.load_refdata" title="rstoolbox.utils.load_refdata"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_refdata</span></code></a>(ref[,&nbsp;homology])</td>
<td>Load the predefined reference data from <code class="docutils literal notranslate"><span class="pre">cath</span></code>, <code class="docutils literal notranslate"><span class="pre">scop</span></code>, <code class="docutils literal notranslate"><span class="pre">scop2</span></code> or <code class="docutils literal notranslate"><span class="pre">chain</span></code>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.utils.make_redundancy_table.html#rstoolbox.utils.make_redundancy_table" title="rstoolbox.utils.make_redundancy_table"><code class="xref py py-obj docutils literal notranslate"><span class="pre">make_redundancy_table</span></code></a>([precalculated,&nbsp;select])</td>
<td>Query into the PDB to retrieve the pre-calculated homology tables.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.plot.plot_in_context.html#rstoolbox.plot.plot_in_context" title="rstoolbox.plot.plot_in_context"><code class="xref py py-obj docutils literal notranslate"><span class="pre">plot_in_context</span></code></a>(df,&nbsp;fig,&nbsp;grid,&nbsp;refdata[,&nbsp;…])</td>
<td>Plot position of decoys in a backgroud reference dataset.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.plot.distribution_quality.html#rstoolbox.plot.distribution_quality" title="rstoolbox.plot.distribution_quality"><code class="xref py py-obj docutils literal notranslate"><span class="pre">distribution_quality</span></code></a>(df,&nbsp;refdata,&nbsp;values,&nbsp;…)</td>
<td>Locate the quantile position of each putative <code class="xref py py-class docutils literal notranslate"><span class="pre">DesingSerie</span></code> in a list of score distributions.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="utils-transforms">
<h2>Utils: Transforms<a class="headerlink" href="#utils-transforms" title="Permalink to this headline">¶</a></h2>
<p>Special functions to help transform your data can be loaded through <code class="docutils literal notranslate"><span class="pre">rstoolbox.utils</span></code>.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.utils.add_column.html#rstoolbox.utils.add_column" title="rstoolbox.utils.add_column"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_column</span></code></a>(df,&nbsp;name,&nbsp;value)</td>
<td>Adds a new column to the DataFrame with the given value.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.utils.split_values.html#rstoolbox.utils.split_values" title="rstoolbox.utils.split_values"><code class="xref py py-obj docutils literal notranslate"><span class="pre">split_values</span></code></a>(df,&nbsp;keys)</td>
<td>Reshape the data to aide plotting of multiple comparable scores.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.utils.split_dataframe_rows.html#rstoolbox.utils.split_dataframe_rows" title="rstoolbox.utils.split_dataframe_rows"><code class="xref py py-obj docutils literal notranslate"><span class="pre">split_dataframe_rows</span></code></a>(df,&nbsp;column_selectors[,&nbsp;…])</td>
<td>Given a dataframe in which certain columns are lists, it splits these lists making new rows in the <a class="reference external" href="http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame" title="(in pandas v0.24.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">DataFrame</span></code></a> out of itself.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.utils.report.html#rstoolbox.utils.report" title="rstoolbox.utils.report"><code class="xref py py-obj docutils literal notranslate"><span class="pre">report</span></code></a>(df)</td>
<td>Cast <strong>basic sequence count</strong> into <strong>pdb count</strong> for the appropiate columns.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.utils.concat_fragments.html#rstoolbox.utils.concat_fragments" title="rstoolbox.utils.concat_fragments"><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat_fragments</span></code></a>(fragment_list)</td>
<td>Combine multiple <a class="reference internal" href="generated/rstoolbox.components.FragmentFrame.html#rstoolbox.components.FragmentFrame" title="rstoolbox.components.FragmentFrame"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentFrame</span></code></a>.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="utils-rosettascript">
<h2>Utils: RosettaScript<a class="headerlink" href="#utils-rosettascript" title="Permalink to this headline">¶</a></h2>
<p>Get the <strong>RosettaScripts</strong> that are called by different functions of the library with <code class="docutils literal notranslate"><span class="pre">rstoolbox.utils</span></code>.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.utils.baseline.html#rstoolbox.utils.baseline" title="rstoolbox.utils.baseline"><code class="xref py py-obj docutils literal notranslate"><span class="pre">baseline</span></code></a>([minimize])</td>
<td>RosettaScript to calculate DSSP secondary structure and phi-psi angles.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.utils.mutations.html#rstoolbox.utils.mutations" title="rstoolbox.utils.mutations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">mutations</span></code></a>([seqID])</td>
<td>RosettaScript to execute a <a class="reference external" href="https://www.rosettacommons.org/docs/latest/rosetta_basics/file_types/resfiles">RESFILE</a>.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="utils-experiments">
<h2>Utils: Experiments<a class="headerlink" href="#utils-experiments" title="Permalink to this headline">¶</a></h2>
<p>Special functions to help obtain data from multiple <strong>Next Generation Sequencing</strong> data.Can be loaded through <code class="docutils literal notranslate"><span class="pre">rstoolbox.utils</span></code>.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.utils.translate_dna_sequence.html#rstoolbox.utils.translate_dna_sequence" title="rstoolbox.utils.translate_dna_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate_dna_sequence</span></code></a>(sequence)</td>
<td>Translates <strong>DNA</strong> to <strong>protein</strong>.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.utils.translate_3frames.html#rstoolbox.utils.translate_3frames" title="rstoolbox.utils.translate_3frames"><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate_3frames</span></code></a>(sequence[,&nbsp;matches])</td>
<td>Translates <strong>DNA</strong> to <strong>protein</strong> trying all possible frames.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.utils.adapt_length.html#rstoolbox.utils.adapt_length" title="rstoolbox.utils.adapt_length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">adapt_length</span></code></a>(seqlist,&nbsp;start,&nbsp;stop[,&nbsp;inclusive])</td>
<td>Pick only the sequence between the provided pattern tags.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.utils.sequencing_enrichment.html#rstoolbox.utils.sequencing_enrichment" title="rstoolbox.utils.sequencing_enrichment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequencing_enrichment</span></code></a>(indata[,&nbsp;enrichment,&nbsp;…])</td>
<td>Retrieve data from multiple <a class="reference external" href="https://www.wikiwand.com/en/DNA_sequencing#/Next-generation_methods">NGS</a> files.</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="internals-functions-for-developers">
<h2>Internals: Functions for Developers<a class="headerlink" href="#internals-functions-for-developers" title="Permalink to this headline">¶</a></h2>
<p>This functions are only of interest if you plan on writing new functionalities in <code class="docutils literal notranslate"><span class="pre">rstoolbox</span></code>.</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.io.open_rosetta_file.html#rstoolbox.io.open_rosetta_file" title="rstoolbox.io.open_rosetta_file"><code class="xref py py-obj docutils literal notranslate"><span class="pre">io.open_rosetta_file</span></code></a>(filename[,&nbsp;multi,&nbsp;…])</td>
<td><em>Internal function</em>; reads through a Rosetta silent file and yields only the lines that the library knows how to parse.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.components.get_selection.html#rstoolbox.components.get_selection" title="rstoolbox.components.get_selection"><code class="xref py py-obj docutils literal notranslate"><span class="pre">components.get_selection</span></code></a>(key_residues,&nbsp;seqID)</td>
<td><em>Internal function</em>; global management and casting of <a class="reference internal" href="generated/rstoolbox.components.Selection.html#rstoolbox.components.Selection" title="rstoolbox.components.Selection"><code class="xref py py-class docutils literal notranslate"><span class="pre">Selection</span></code></a>.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.utils.make_rosetta_app_path.html#rstoolbox.utils.make_rosetta_app_path" title="rstoolbox.utils.make_rosetta_app_path"><code class="xref py py-obj docutils literal notranslate"><span class="pre">utils.make_rosetta_app_path</span></code></a>(application)</td>
<td>Provided the expected Rosetta application, add path and suffix.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.tests.helper.random_frequency_matrix.html#rstoolbox.tests.helper.random_frequency_matrix" title="rstoolbox.tests.helper.random_frequency_matrix"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tests.helper.random_frequency_matrix</span></code></a>(size[,&nbsp;…])</td>
<td>Generate a random frequency matrix.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="generated/rstoolbox.tests.helper.random_proteins.html#rstoolbox.tests.helper.random_proteins" title="rstoolbox.tests.helper.random_proteins"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tests.helper.random_proteins</span></code></a>(size,&nbsp;count)</td>
<td>Generate random protein sequences.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="generated/rstoolbox.tests.helper.random_fastq.html#rstoolbox.tests.helper.random_fastq" title="rstoolbox.tests.helper.random_fastq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">tests.helper.random_fastq</span></code></a>(sequence,&nbsp;…)</td>
<td>Generate a requested number of fastq files.</td>
</tr>
</tbody>
</table>
</div>
</div>


    </div>
      
  </div>
</div>

<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
        <br/>
        
      
    </p>
    <p>
        &copy; Copyright 2018, Jaume Bonet.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.2.<br/>
    </p>
  </div>
</footer>
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-74877325-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-74877325-2');
</script>

  </body>
</html>